stdlib/sql/to.go

```diff
@@ -0,0 +1,361 @@
+	spec  *ToSQLProcedureSpec
+	db    *sql.DB
+	tx    *sql.Tx
+}
+
+func (t *ToSQLTransformation) RetractTable(id execute.DatasetID, key flux.GroupKey) error {
+	return t.d.RetractTable(key)
+}
+
+func NewToSQLTransformation(d execute.Dataset, cache execute.TableBuilderCache, spec *ToSQLProcedureSpec) *ToSQLTransformation {
+	db, err := sql.Open(spec.Spec.DriverName, spec.Spec.DataSourceName)
+	if err != nil {
+		panic(err)
+	}
+	var tx *sql.Tx
+	if spec.Spec.DriverName != "sqlmock" {
+		tx, err = db.Begin()
+		if err != nil {
+			panic(err)
+		}
+	}
+	return &ToSQLTransformation{
+		d:     d,
+		cache: cache,
+		spec:  spec,
+		db:    db,
+		tx:    tx,
+	}
+}
+
+type idxType struct {
+	Idx  int
+	Type flux.ColType
+}
+
+func (t *ToSQLTransformation) Process(id execute.DatasetID, tbl flux.Table) (err error) {
+	cols := tbl.Cols()
+	labels := make(map[string]idxType, len(cols))
+	var colNames, questionMarks, newSQLTableCols []string
+	for i, col := range cols {
+		labels[col.Label] = idxType{Idx: i, Type: col.Type}
+		questionMarks = append(questionMarks, "?")
+		colNames = append(colNames, col.Label)
+
+		switch col.Type {
+		case flux.TFloat:
+			newSQLTableCols = append(newSQLTableCols, fmt.Sprintf("%s FLOAT", col.Label))
+		case flux.TInt:
+			newSQLTableCols = append(newSQLTableCols, fmt.Sprintf("%s BIGINT", col.Label))
+		case flux.TUInt:
+			newSQLTableCols = append(newSQLTableCols, fmt.Sprintf("%s BIGINT", col.Label))
+		case flux.TString:
+			switch t.spec.Spec.DriverName {
+			case "mysql":
+				newSQLTableCols = append(newSQLTableCols, fmt.Sprintf("%s TEXT(16383)", col.Label))
+			case "postgres":
+				newSQLTableCols = append(newSQLTableCols, fmt.Sprintf("%s text", col.Label))
+			}
+		case flux.TTime:
+			newSQLTableCols = append(newSQLTableCols, fmt.Sprintf("%s DATETIME", col.Label))
+		case flux.TBool:
+			newSQLTableCols = append(newSQLTableCols, fmt.Sprintf("%s BOOL", col.Label))
+		default:
+			return fmt.Errorf("invalid type for column %s", col.Label)
+		}
+	}
+
+	// Creates the placeholders for values in the query
+	// eg: (?,?)
+	valuePlaceHolders := fmt.Sprintf("(%s)", strings.Join(questionMarks, ","))
+
+	builder, new := t.cache.TableBuilder(tbl.Key())
+	if new {
+		if err := execute.AddTableCols(tbl, builder); err != nil {
+			return err
+		}
+	}
+
+	var wg syncutil.WaitGroup
+	wg.Do(func() error {
+		err := tbl.Do(func(er flux.ColReader) error {
+			l := er.Len()
+
+			// valueStrings is an array of valuePlaceHolders, which will be joined later
+			valueStrings := make([]string, 0, l)
+			// valueArgs holds all the values to pass into the query
+			valueArgs := make([]interface{}, 0, l*len(cols))
+
+			if t.spec.Spec.DriverName != "sqlmock" {
+				q := fmt.Sprintf("CREATE TABLE IF NOT EXISTS %s (%s)", t.spec.Spec.Table, strings.Join(newSQLTableCols, ","))
+				_, err = t.tx.Exec(q)
+				if err != nil {
+					if rbErr := t.tx.Rollback(); err != nil {
+						return fmt.Errorf("\n%s\n%s", err, rbErr)
```

Comment:
new lines in errors cause us problems.   make this "transaction failed (%s) while recovering from %s"